### PR TITLE
Add fx.Decorate

### PR DESCRIPTION
**Pull request by sywhang**
This adds `fx.Decorate`, which lets you specify decorators to an fx app. A decorator can take in one or more dependencies that have already been `Provide`d to the app, and produce one or more values that will be used as replacements in the object graph.

For example, suppose there is a simple app like this:
```go

fx.New(
  fx.Provide(func() *Logger {
    return &Logger{Name: "logger"}
  }),
  fx.Invoke(func(l *Logger) {
    fmt.Println(l.Name)
  }),
)
```

Running this app will print "logger" on the console.

Now let us suppose a decorator was provided:
```go
fx.New(
  fx.Provide(...), // Provide same function as above
  fx.Decorate(func(l *Logger) *Logger {
    return &Logger{Name: "decorated " + l.Name}
  }),
  fx.Invoke(...), // Invoke same function as above
)
```

The decorator here will take in the provided Logger and replace it with another logger whose `Name` is `decorated logger`. The `Invoke`d function is then executed with this replacement value, so running this app will print "decorated logger" on the console.

In terms of implementation, a decorator is represented by the target decorator function and the call stack it was provided from, similar to a provider. `module` contains a list of decorators that were specified within its scope.

The dig dependency had to be updated to the latest master branch of Dig to ensure the fix for https://github.com/uber-go/dig/issues/316 is in.

Following this PR, there are two additional pieces I will be adding:
1. An eventing system for fx.Decorate. 
2. fx.Replace, which takes in a value instead of a function to replace a value in the object graph. This is similar to what fx.Supply is to fx.Provide.

This PR along with the two PRs above should make the long-awaited feature of graph modifications in fx finally possible.

---

Refs #653, #649, #825, uber-go/dig#230, GO-1203, GO-736

### CODE REVIEW

File: decorate.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"go.uber.org/dig"
+	"go.uber.org/fx/internal/fxreflect"
+)
+
+// Decorate specifies one or more decorator functions to an fx application.
```
Comment (abhinav) [inline decorate.go:31]:
```suggestion
// Decorate specifies one or more decorator functions to an Fx application.
```

File: decorate.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+//
+// The following decorator accepts multiple dependencies from the graph, augments and returns
+// one of them.
+//
+//  fx.Decorate(func(log *zap.Logger, cfg *Config) *zap.Logger {
+//    return log.Named(cfg.Name)
+//  })
+//
+// All modifications in the object graph due to a decorator are scoped to the fx.Module it was
+// specified from.
```
Comment (abhinav) [inline decorate.go:51]: Worth adding:

> Decorations specified in the top-level fx.New call apply across the application.

(We can add more, like decoration chaining, value group decoration, etc. later)

File: decorate.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+//    return log.Named("myapp")
+//  })
+//
+// The following decorator accepts multiple dependencies from the graph, augments and returns
+// one of them.
+//
+//  fx.Decorate(func(log *zap.Logger, cfg *Config) *zap.Logger {
+//    return log.Named(cfg.Name)
+//  })
+//
```
Comment (abhinav) [inline decorate.go:49]: Worth adding:

> Similar to fx.Provide, functions passed to fx.Decorate may optionally return an error as their last result. If a decorator returns a non-nil error, it will halt application startup.

File: decorate.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	items := make([]string, len(o.Targets))
+	for i, f := range o.Targets {
+		items[i] = fxreflect.FuncName(f)
+	}
+	return fmt.Sprintf("fx.Decorate(%s)", strings.Join(items, ", "))
+}
+
+// provide is a single decorators used in Fx.
+type decorator struct {
+	// Constructor provided to Fx. This may be an fx.Annotated.
```
Comment (abhinav) [inline decorate.go:83]: outdated: this may not be an fx.Annotated. This may be an fx.annotated, produced by fx.Annotate.

(Can probably document separately that decorators support fx.Annotate, but not fx.Annotated.)

File: decorate_test.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+			fx.Decorate(func() *Logger {
+				return &Logger{Name: "testRedis"}
+			}),
+			fx.Invoke(func(l *Logger) {
+				assert.Equal(t, "testRedis", l.Name)
+			}),
+		)
+
+		app := fxtest.New(t,
+			testRedis,
```
Comment (abhinav) [inline decorate_test.go:56]: This is fine as a test, but in the real world, I expect we will not nest "redis" under "testredis".

File: decorate_test.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+			fx.Invoke(func(l *Logger) {
+				assert.Equal(t, "redis", l.Name)
+			}),
+		)
+		defer app.RequireStart().RequireStop()
+	})
+
+	t.Run("decorate a dependency from root", func(t *testing.T) {
```
Comment (abhinav) [inline decorate_test.go:64]: description does not match test

File: decorate_test.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+			redis,
+			fx.Provide(func() *Logger {
+				assert.Fail(t, "should not run this")
+				return &Logger{Name: "root"}
+			}),
+		)
+		defer app.RequireStart().RequireStop()
+	})
+
+	t.Run("use Decorate in root", func(t *testing.T) {
```
Comment (abhinav) [inline decorate_test.go:83]: nice

File: decorate_test.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+		takeout := fx.Module("takeout",
+			cafe,
+			fx.Decorate(fx.Annotate(func(coffee []*Coffee) []*Coffee {
+				var newC []*Coffee
+				for _, c := range coffee {
+					newC = append(newC, &Coffee{
+						Name:  c.Name,
+						Price: c.Price + 1,
+					})
```
Comment (abhinav) [inline decorate_test.go:130]: 😆

File: module.go [hunk 9]
```diff
@@ -174,3 +176,17 @@ func (m *module) executeInvoke(i invoke) (err error) {
...
+
+func (m *module) decorate() (err error) {
+	for _, decorator := range m.decorators {
+		if err := runDecorator(m.scope, decorator); err != nil {
+			return err
+		}
+	}
+	for _, m := range m.modules {
+		if err := m.decorate(); err != nil {
+			return err
```
Comment (abhinav) [inline module.go:188]: failure in a decorator inside a nested module should cover this, I think.

File: decorate.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+func (o decorateOption) apply(mod *module) {
+	for _, target := range o.Targets {
+		mod.decorators = append(mod.decorators, decorator{
+			Target: target,
+			Stack:  o.Stack,
+		})
+	}
+}
+
+func (o decorateOption) String() string {
```
Comment (abhinav) [inline decorate.go:73]: Let's test this one too. there's a string test for all the other options, I believe.

**Review by abhinav — APPROVED** [review]
Surprisingly simple after all of the other changes. No major concerns.
Minus the couple missing untested corner cases, I think this is good to go.

Please get a review from someone else on the team as well.

---

As an aside: 🎉 🎉 🎉

File: decorate_test.go [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+				assert.Fail(t, "this should never run")
+			}),
+		)
+
+		err := app.Err()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "missing dependencies")
+	})
+
+	t.Run("use a decorator like a provider", func(t *testing.T) {
```
Comment (r-hang) [inline decorate_test.go:335]:
```suggestion
	t.Run("decorate cannot provide a non-existent type", func(t *testing.T) {
```

File: decorate_test.go [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+				assert.Fail(t, "this should never run")
+			}, fx.ParamTags(tag))),
+		)
+
+		err := app.Err()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "major sadness")
+	})
+
+	t.Run("decorator missing a dependency", func(t *testing.T) {
```
Comment (r-hang) [inline decorate_test.go:310]:
```suggestion
	t.Run("all decorator dependencies must be provided", func(t *testing.T) {
```

File: decorate_test.go [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+				}),
+			),
+		)
+
+		err := app.Err()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "minor sadness")
+	})
+
+	t.Run("decorate the same type twice from the same Module", func(t *testing.T) {
```
Comment (r-hang) [inline decorate_test.go:265]:
```suggestion
	t.Run("decorating a type more than once in the same Module errors", func(t *testing.T) {
```

**Review by r-hang — APPROVED** [review]

**Comment by akshayjshah** [comment]
Heck yeah! Nice work :)

**Comment by Ardagan** [comment]
Great work!
Is there any rough estimation when this will end up in released version of fx?

**Comment by sywhang** [comment]
Thanks @Ardagan! Our current estimate is to ship this by end of the month. We have some more work to do to integrate this with the fxevent package, and add fx.Replace (though there's already a PR out for that - #837). We also need to verify this doesn't break any Go services here at Uber.

After all those have been done, we should be good to go!

**Comment by sywhang** [comment]
@Ardagan it's been released now: https://github.com/uber-go/fx/releases/tag/v1.17.0
